workflow/sync/sync_manager.go

```diff
@@ -272,11 +272,14 @@ func (cm *Manager) ReleaseAll(wf *wfv1.Workflow) bool {
 }
 
 func ensureInit(wf *wfv1.Workflow, lockType wfv1.SynchronizationType) {
-	if lockType == wfv1.SynchronizationTypeSemaphore && (wf.Status.Synchronization == nil || wf.Status.Synchronization.Semaphore == nil) {
```

Comment:
If workflow has mutex and semaphore and mutext initialized `wf.Status.Synchronization` before. This code was re-initializing `wf.Status.Synchronization` with only semaphore. So mutex is nil which is causing nil pointer expection.